### ISSUE TITLE
Add support for arrays and tuples of custom types

### ISSUE THREAD

**Issue by digorithm** [FuelLabs/fuels-rs#306]
Let the following Sway contract:

```Rust
contract;

struct Person {
    name: str[4]
}

abi TestContract {
    fn return_array_of_structs(param: [Person; 2]) -> [Person; 2];
}


impl TestContract for Contract {
    fn return_array_of_structs(param: [Person; 2]) -> [Person; 2] {
        param
    }
}
```

We currently don't seem to support that, the same applies to tuples containing custom types. This support is needed in https://github.com/FuelLabs/fuels-rs/issues/296. However, the compiler is also failing at outputting the proper JSON ABI in this case, see [here](https://github.com/FuelLabs/sway/issues/1631). 

Once https://github.com/FuelLabs/sway/issues/1631 is done, this should be implemented here as well.

**Comment by adlerjohn** [comment]
Wait, this isn't yet implemented in fuels-rs, no?

**Comment by digorithm** [comment]
> Wait, this isn't yet implemented in fuels-rs, no?

That's right. We still need to implement on the SDK side. It was just blocked by the compiler. Which isn't the case anymore.  

The issue should stay open.

**Comment by mohammadfawaz** [comment]
@digorithm we should probably update this issue to also handle tuples of custom types (Ref https://github.com/FuelLabs/sway/pull/1653) ? Or would you like that to be tracked in a separate issue?

**Comment by digorithm** [comment]
Yeah, bundling these together in this issue makes sense. I'll be touching the same sections of the codebase.
